File: src/neml2/models/kwn/RateLimitedPrecipitateGrowthRate.cxx
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "neml2/models/kwn/RateLimitedPrecipitateGrowthRate.h"

#include "neml2/tensors/Scalar.h"
#include "neml2/tensors/functions/diagonalize.h"
#include "neml2/tensors/functions/imap.h"

namespace neml2
{
register_NEML2_object(RateLimitedPrecipitateGrowthRate);

OptionSet
RateLimitedPrecipitateGrowthRate::expected_options()
{
  OptionSet options = Model::expected_options();
  options.doc() = "Compute the rate-limited precipitate growth rate for a single species.";

  options.set_input("radius");
  options.set("radius").doc() = "Precipitate radius per size bin";

  options.set_input("current_concentration");
  options.set("current_concentration").doc() = "Current concentration in solution";

  options.set_parameter<TensorName<Scalar>>("equilibrium_concentration");
  options.set("equilibrium_concentration").doc() = "Equilibrium concentration in solution";

  options.set_parameter<TensorName<Scalar>>("precipitate_concentration");
  options.set("precipitate_concentration").doc() = "Precipitate concentration";

  options.set_parameter<TensorName<Scalar>>("diffusivity");
  options.set("diffusivity").doc() = "Species diffusivity in solution";

  options.set_output("growth_rate");
  options.set("growth_rate").doc() = "Precipitate growth rate per size bin";

  return options;
}

RateLimitedPrecipitateGrowthRate::RateLimitedPrecipitateGrowthRate(const OptionSet & options)
  : Model(options),
    _R(declare_input_variable<Scalar>("radius")),
    _x(declare_input_variable<Scalar>("current_concentration")),
    _x_eq(declare_parameter<Scalar>("x_eq", "equilibrium_concentration", true)),
    _x_p(declare_parameter<Scalar>("x_p", "precipitate_concentration", true)),
    _D(declare_parameter<Scalar>("D", "diffusivity", true)),
    _R_dot(declare_output_variable<Scalar>("growth_rate"))
{
}

void
RateLimitedPrecipitateGrowthRate::set_value(bool out, bool dout_din, bool /*d2out_din2*/)
{
  const auto nbin = _R.intmd_size(-1);
  const auto R = _R();
  const auto x_inf = (_x.intmd_dim() == 0 ? _x().intmd_expand(nbin) : _x());
  const auto x_eq = (_x_eq.intmd_dim() == 0 ? _x_eq.intmd_expand(nbin) : _x_eq);
  const auto x_p = (_x_p.intmd_dim() == 0 ? _x_p.intmd_expand(nbin) : _x_p);
  const auto D = (_D.intmd_dim() == 0 ? _D.intmd_expand(nbin) : _D);

  const auto denom = x_p - x_eq;
  const auto numer = x_inf - x_eq;
  const auto coef = D / R;

  if (out)
    _R_dot = coef * numer / denom;

  if (dout_din)
  {
    const auto denom2 = denom * denom;

    if (_R.is_dependent())
    {
      const auto d_rate_dR = -coef * numer / (denom * R);
      const auto r_map = imap_v<Scalar>(_R.options()).intmd_expand(nbin);
      const auto diag_r = intmd_diagonalize(r_map);
      _R_dot.d(_R, 2, 1, 1) = d_rate_dR.intmd_unsqueeze(1) * diag_r;
    }

    if (_x.is_dependent())
    {
      const auto d_rate_dx = coef / denom;
      if (_x.intmd_dim() == 0)
        _R_dot.d(_x, 1, 1, 0) = d_rate_dx;
      else
      {
        const auto x_map = imap_v<Scalar>(_x.options()).intmd_expand(nbin);
        const auto diag_x = intmd_diagonalize(x_map);
        _R_dot.d(_x, 2, 1, 1) = d_rate_dx.intmd_unsqueeze(1) * diag_x;
      }
    }

    if (const auto * const x_eq_param = nl_param("x_eq"))
    {
      const auto d_rate_dx_eq = D * (x_inf - x_p) / (R * denom2);
      if (x_eq_param->intmd_dim() == 0)
        _R_dot.d(*x_eq_param, 1, 1, 0) = d_rate_dx_eq;
      else
      {
        const auto xeq_map = imap_v<Scalar>(_x_eq.options()).intmd_expand(nbin);
        const auto diag_xeq = intmd_diagonalize(xeq_map);
        _R_dot.d(*x_eq_param, 2, 1, 1) = d_rate_dx_eq.intmd_unsqueeze(1) * diag_xeq;
      }
    }

    if (const auto * const x_p_param = nl_param("x_p"))
    {
      const auto d_rate_dx_p = -D * numer / (R * denom2);
      if (x_p_param->intmd_dim() == 0)
        _R_dot.d(*x_p_param, 1, 1, 0) = d_rate_dx_p;
      else
      {
        const auto xp_map = imap_v<Scalar>(_x_p.options()).intmd_expand(nbin);
        const auto diag_xp = intmd_diagonalize(xp_map);
        _R_dot.d(*x_p_param, 2, 1, 1) = d_rate_dx_p.intmd_unsqueeze(1) * diag_xp;
      }
    }

    if (const auto * const D_param = nl_param("D"))
    {
      const auto d_rate_dD = numer / (R * denom);
      if (D_param->intmd_dim() == 0)
        _R_dot.d(*D_param, 1, 1, 0) = d_rate_dD;
      else
      {
        const auto d_map = imap_v<Scalar>(_D.options()).intmd_expand(nbin);
        const auto diag_d = intmd_diagonalize(d_map);
        _R_dot.d(*D_param, 2, 1, 1) = d_rate_dD.intmd_unsqueeze(1) * diag_d;
      }
    }
  }
}
} // namespace neml2

File: neml2/models/kwn/RateLimitedPrecipitateGrowthRate.py
# Copyright 2024, UChicago Argonne, LLC
# All Rights Reserved
# Software Name: NEML2 -- the New Engineering material Model Library, version 2
# By: Argonne National Laboratory
# OPEN SOURCE LICENSE (MIT)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

"""Python-native mirror of the C++ ``RateLimitedPrecipitateGrowthRate`` model."""

from __future__ import annotations

from ...chain_rule import ChainRuleAction, ChainRuleDict
from ...factory import register_native
from ...model import Model
from ...schema import HitSchema, input, output, parameter
from ...types import Scalar


@register_native("RateLimitedPrecipitateGrowthRate")
class RateLimitedPrecipitateGrowthRate(Model):
    """Compute the rate-limited precipitate growth rate for a single species."""

    hit = HitSchema(
        input("current_concentration", Scalar, "Current concentration in solution"),
        output("growth_rate", Scalar, "Precipitate growth rate per size bin"),
        parameter(
            "radius",
            Scalar,
            "Precipitate radius per size bin",
            attr="R",
            allow_nonlinear=True,
        ),
        parameter(
            "equilibrium_concentration",
            Scalar,
            "Equilibrium concentration in solution",
            attr="x_eq",
            allow_nonlinear=True,
        ),
        parameter(
            "concentration_difference",
            Scalar,
            "Concentration difference between precipitate and equilibrium",
            attr="dx",
            allow_nonlinear=True,
        ),
        parameter(
            "diffusivity",
            Scalar,
            "Species diffusivity in solution",
            attr="D",
            allow_nonlinear=True,
        ),
    )
    # ``radius`` is a per-bin (sub_batch_ndim=1) Scalar while the other
    # parameters and the structural input may be scalar (sub_batch_ndim=0). The
    # bin-broadcast made by ``align_sub_batch`` makes site-i of ``growth_rate``
    # depend on the single scalar value of any sub_batch_ndim=0 input/promoted
    # parameter -- a cross-bin (dense) coupling rather than the default
    # per-site diagonal one. Declare every input pair as ``"dense"`` so the
    # composed-model resolver picks the conservative (BLOCK -> DENSE fallback)
    # sparsity flag. Per-bin inputs still produce a strictly bin-diagonal
    # action; the flag is purely an upper bound used by the system assembler.
    list_deriv = {
        ("growth_rate", "current_concentration"): "dense",
        ("growth_rate", "radius"): "dense",
        ("growth_rate", "equilibrium_concentration"): "dense",
        ("growth_rate", "concentration_difference"): "dense",
        ("growth_rate", "diffusivity"): "dense",
    }

    # ``from_hit`` auto-declares the four parameters (stored as R, x_eq, dx, D).
    # Annotate so pyright sees the typed wrappers that ``Model.__getattr__``
    # returns; runtime resolution is via ``_get_param`` so the same code path
    # handles both static and nl-promoted values.
    R: Scalar
    x_eq: Scalar
    dx: Scalar
    D: Scalar

    def forward(  # type: ignore[override]
        self,
        x: Scalar,
        *nl_params: Scalar,
        v: ChainRuleDict | None = None,
    ) -> Scalar | tuple[Scalar, ChainRuleDict]:
        R = self._get_param("R", nl_params, Scalar)
        x_eq = self._get_param("x_eq", nl_params, Scalar)
        dx = self._get_param("dx", nl_params, Scalar)
        D = self._get_param("D", nl_params, Scalar)

        # Forward: R_dot = (D / R) * (x - x_eq) / dx. Typed Scalar algebra
        # end-to-end; align_sub_batch transparently lifts sub_batch_ndim=0
        # parameters (or input) up to the per-bin sub_batch_ndim of R, matching
        # the C++ ``intmd_expand(nbin)`` calls in ``set_value``.
        numer = x - x_eq
        coef = D / R
        rate = coef * numer / dx

        if v is None:
            return rate

        # Differential pushforward. The forward is a simple rational
        # function of (x, R, x_eq, dx, D); every action is a closed-form
        # coefficient times the input tangent. Wrapper algebra handles the
        # scalar-vs-per-bin broadcast on the tangent the same way it did on
        # the forward, so each action stays pure typed Scalar arithmetic -- no
        # diagonal embedding, no Jacobian materialization.
        #
        #   d R_dot / d x    = D / (R * dx)        =  coef / dx
        #   d R_dot / d x_eq = -D / (R * dx)       = -coef / dx
        #   d R_dot / d dx   = -D * numer / (R * dx^2)
        #   d R_dot / d R    = -D * numer / (R^2 * dx)
        #   d R_dot / d D    = numer / (R * dx)
        actions: dict[str, ChainRuleAction] = {}

        d_rate_dx = coef / dx

        def x_action(V: Scalar) -> Scalar:
            return d_rate_dx * V

        actions["current_concentration"] = x_action

        x_eq_nlp = self._nl_params.get("x_eq")
        if x_eq_nlp is not None:
            d_rate_dxeq = -coef / dx

            def x_eq_action(V: Scalar) -> Scalar:
                return d_rate_dxeq * V

            actions[x_eq_nlp.input_name] = x_eq_action

        dx_nlp = self._nl_params.get("dx")
        if dx_nlp is not None:
            d_rate_ddx = -coef * numer / (dx * dx)

            def dx_action(V: Scalar) -> Scalar:
                return d_rate_ddx * V

            actions[dx_nlp.input_name] = dx_action

        R_nlp = self._nl_params.get("R")
        if R_nlp is not None:
            d_rate_dR = -coef * numer / (dx * R)

            def R_action(V: Scalar) -> Scalar:
                return d_rate_dR * V

            actions[R_nlp.input_name] = R_action

        D_nlp = self._nl_params.get("D")
        if D_nlp is not None:
            d_rate_dD = numer / (R * dx)

            def D_action(V: Scalar) -> Scalar:
                return d_rate_dD * V

            actions[D_nlp.input_name] = D_action

        return rate, self.apply_chain_rule(v, "growth_rate", actions, output=rate)


__all__ = ["RateLimitedPrecipitateGrowthRate"]
